test/javascripts/turbolinks_test.coffee

```diff
@@ -1,210 +1,946 @@
+    """
+
+  urlFor = (slug) ->
     window.location.origin + slug
 
+  visit = ({options, url}, callback) ->
+    $(document).one('page:load', (event) ->
+      setTimeout((-> callback(event)), 0)
+    )
+    Turbolinks.visit('/' + url, options)
+
   beforeEach ->
-    @server = sinon.fakeServer.create()
+    pushStateStub = stub(Turbolinks, 'pushState')
+    replaceStateStub = stub(Turbolinks, 'replaceState')
+    stub(Turbolinks, 'fullPageNavigate', -> $(document).trigger('page:load'))
+
+    server = sinon.fakeServer.create()
+    Object.keys(ROUTES).forEach (url) ->
+      server.respondWith('/' + url, ROUTES[url])
+    server.autoRespond = true
+
     $("script").attr("data-turbolinks-eval", false)
     $("#mocha").attr("refresh-never", true)
-    @replaceStateStub = stub(Turbolinks, "replaceState")
-    @pushStateStub = stub(Turbolinks, "pushState")
-    document.body.appendChild(createTurboNodule())
+
+    resetPage()
 
   afterEach ->
-    @server.restore()
-    @pushStateStub.restore()
-    @replaceStateStub.restore()
-    $("#turbo-area").remove()
+    Turbolinks.fullPageNavigate.restore()
+    Turbolinks.pushState.restore()
+    Turbolinks.replaceState.restore()
 
-  html_one = """
-    <!doctype html>
-    <html>
-      <head>
-        <title>Hi there!</title>
-      </head>
-      <body>
-        <div>YOLO</div>
-        <div id="turbo-area" refresh="turbo-area">Hi bob</div>
-      </body>
-    </html>
-  """
-
-  script_response = """
-    <!doctype html>
-    <html>
-      <head>
-        <title>Hi</title>
-      </head>
-      <body>
-        <script id="turbo-area" refresh="turbo-area">globalStub()</script>
-      </body>
-    </html>
-  """
-
-  script_response_turbolinks_eval_false = """
-    <!doctype html>
-    <html>
-      <head>
-        <title>Hi</title>
-      </head>
-      <body>
-        <script data-turbolinks-eval="false" id="turbo-area" refresh="turbo-area">globalStub()</script>
-      </body>
-    </html>
-  """
-
-  response_with_refresh_always = """
-    <!doctype html>
-    <html>
-      <head>
-        <title>Hi</title>
-      </head>
-      <body>
-        <div id="div1" refresh="div1">
-          <div id="div2" refresh-always>Refresh-always</div>
-        </div>
-      </body>
-    </html>
-  """
+    $(document).off(TURBO_EVENTS.join(' '))
+    $("#turbo-area").remove()
+    server.restore()
 
   it 'is defined', ->
-    assert Turbolinks
+    assert(Turbolinks)
 
   describe '#visit', ->
-    it 'returns if pageChangePrevented', ->
-      listener = (event) ->
+    it 'only loads once if called multiple times before an XHR finishes', (done) ->
+      pageReceive = stub()
+      $(document).on('page:receive', pageReceive)
+      visit url: 'noScriptsOrLinkInHead', -> true
+      visit url: 'noScriptsOrLinkInHead', ->
+        assert(pageReceive.calledOnce, 'page:receive should only be emitted once!')
+        done()
+
+    it 'returns if pageChangePrevented', (done) ->
```

Comment:
Ideally, this'd be more meaningful.  e.g., `it 'allows listeners to veto page changes'`.  

You didn't pick these test names, I know :(